--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -94,7 +_,7 @@
     public final int entityBroadcastRangePercentage = this.get("entity-broadcast-range-percentage", property -> Mth.clamp(property, 10, 1000), 100);
     public final String textFilteringConfig = this.get("text-filtering-config", "");
     public final int textFilteringVersion = this.get("text-filtering-version", 0);
-    public final Optional<MinecraftServer.ServerResourcePackInfo> serverResourcePackInfo;
+    public Optional<MinecraftServer.ServerResourcePackInfo> serverResourcePackInfo;
     public final DataPackConfig initialDataPackConfiguration;
     public final Settings<DedicatedServerProperties>.MutableValue<Integer> playerIdleTimeout = this.getMutable("player-idle-timeout", 0);
     public final Settings<DedicatedServerProperties>.MutableValue<Boolean> whiteList = this.getMutable("white-list", false);
