From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sat, 20 Jan 2024 16:51:20 +0100
Subject: [PATCH] Ignoring errors when creating javadocs


diff --git a/build.gradle.kts b/build.gradle.kts
index e29e5024fa693baae469d47fe77b57118f14627c..7c8dc021156007448284eefc25e2ae81b21a2a26 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -164,6 +164,7 @@ tasks.withType<Javadoc> {
     options.overview = "src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
+    isFailOnError = false;
     options.links(
         "https://guava.dev/releases/32.1.2-jre/api/docs/",
         "https://javadoc.io/doc/org.yaml/snakeyaml/2.2/",
