From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <lukasjonsson@outlook.de>
Date: Thu, 23 Nov 2023 00:41:02 +0100
Subject: [PATCH] Adding Utility class for MCCreativeLab


diff --git a/src/main/java/de/verdox/mccreativelab/CreativeLabImpl.java b/src/main/java/de/verdox/mccreativelab/CreativeLabImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..f25d35d7d35f8394e5012d2f76c07113c92e7e3f
--- /dev/null
+++ b/src/main/java/de/verdox/mccreativelab/CreativeLabImpl.java
@@ -0,0 +1,7 @@
+package de.verdox.mccreativelab;
+
+import org.jetbrains.annotations.NotNull;
+
+public final class CreativeLabImpl implements CreativeLab {
+
+}
diff --git a/src/main/java/net/minecraft/server/Bootstrap.java b/src/main/java/net/minecraft/server/Bootstrap.java
index ffca3a7df1c5b344dbd77e1db997c76d22a31021..a1db7ce8a87a58e67d540a65d8b88e712f76e94c 100644
--- a/src/main/java/net/minecraft/server/Bootstrap.java
+++ b/src/main/java/net/minecraft/server/Bootstrap.java
@@ -63,6 +63,11 @@ public class Bootstrap {
             Bootstrap.isBootstrapped = true;
             Instant instant = Instant.now();
 
+            // MCCLab - start
+            de.verdox.mccreativelab.MCCreativeLab.setCreativeLab(new de.verdox.mccreativelab.CreativeLabImpl());
+            MinecraftServer.LOGGER.info("MCCreativeLab: initialized");
+            // MCCLab - end
+
             io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enterBootstrappers(); // Paper - Entrypoint for bootstrapping
             if (BuiltInRegistries.REGISTRY.keySet().isEmpty()) {
                 throw new IllegalStateException("Unable to load registries");
