From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Wed, 29 May 2024 15:11:34 +0000
Subject: [PATCH] Add FurnaceRunOutOfFuelEvent


diff --git a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index 15e0861486a2bda3e2f4049b1b5a299c870acd31..9c89397a838d904f45b735c903d8ed16c29363aa 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -306,6 +306,8 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
 
         if (flag1) {
             setChanged(world, pos, state);
+            if(!flag)
+                new org.bukkit.event.inventory.FurnaceRunOutOfFuelEvent(CraftBlock.at(world, pos)).callEvent();
         }
 
     }
