From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Thu, 20 Jun 2024 12:39:36 +0000
Subject: [PATCH] Removing enums


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index 132afec6bceb6c866de0aeb83db9613d4e74e2e7..fa630358bba4df12cce28a35c091562f0e361864 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -99,6 +99,11 @@ public final class PaperRegistries {
             entry(Registries.SOUND_EVENT, RegistryKey.SOUND_EVENT, Sound.class, CraftSound::new),
             entry(Registries.DATA_COMPONENT_TYPE, RegistryKey.DATA_COMPONENT_TYPE, DataComponentType.class, PaperDataComponentType::of),
 
+            entry(Registries.ATTRIBUTE, RegistryKey.ATTRIBUTE, org.bukkit.attribute.Attribute.class, org.bukkit.craftbukkit.attribute.CraftAttribute::new),
+            entry(Registries.POINT_OF_INTEREST_TYPE, RegistryKey.POI_TYPE, org.bukkit.PoiType.class, org.bukkit.craftbukkit.CraftPoiType::new),
+            entry(Registries.SCHEDULE, RegistryKey.ENTITY_SCHEDULE, org.bukkit.EntitySchedule.class, org.bukkit.craftbukkit.CraftEntitySchedule::new),
+            entry(Registries.ACTIVITY, RegistryKey.ENTITY_ACTIVITY, org.bukkit.EntityActivity.class, org.bukkit.craftbukkit.CraftEntityActivity::new),
+
             // data-drivens
             entry(Registries.BIOME, RegistryKey.BIOME, Biome.class, CraftBiome::new).delayed(),
             entry(Registries.STRUCTURE, RegistryKey.STRUCTURE, Structure.class, CraftStructure::new).delayed(),
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftEntityActivity.java b/src/main/java/org/bukkit/craftbukkit/CraftEntityActivity.java
new file mode 100644
index 0000000000000000000000000000000000000000..776a4b14e0a5d78ef183c1c012d7ffcae8b59876
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/CraftEntityActivity.java
@@ -0,0 +1,33 @@
+package org.bukkit.craftbukkit;
+
+import net.minecraft.core.registries.Registries;
+import org.bukkit.EntityActivity;
+import net.minecraft.world.entity.schedule.Activity;
+import org.bukkit.NamespacedKey;
+import org.bukkit.PoiType;
+import org.bukkit.Registry;
+import org.jetbrains.annotations.NotNull;
+
+public class CraftEntityActivity extends BukkitToNMSMapping<Activity> implements EntityActivity {
+    public static EntityActivity minecraftToBukkit(Activity minecraft) {
+        return minecraftToBukkit(minecraft, Registries.ACTIVITY, Registry.ENTITY_ACTIVITY);
+    }
+
+    public static Activity bukkitToMinecraft(EntityActivity bukkit) {
+        return bukkitToMinecraft(bukkit, Registries.ACTIVITY);
+    }
+
+    public CraftEntityActivity(NamespacedKey key, Activity handle) {
+        super(key, handle);
+    }
+
+    @Override
+    public @NotNull String translationKey() {
+        return "activity.minecraft." + getKey().getKey();
+    }
+
+    @Override
+    public String toString() {
+        return "CraftEntityActivity{key=" + this.getKey() + "}";
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftEntitySchedule.java b/src/main/java/org/bukkit/craftbukkit/CraftEntitySchedule.java
new file mode 100644
index 0000000000000000000000000000000000000000..aed77f45b42aeff11e38c87226c2ac198c86d837
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/CraftEntitySchedule.java
@@ -0,0 +1,34 @@
+package org.bukkit.craftbukkit;
+
+import net.minecraft.core.registries.Registries;
+import net.minecraft.world.entity.schedule.Activity;
+import net.minecraft.world.entity.schedule.Schedule;
+import org.bukkit.EntityActivity;
+import org.bukkit.EntitySchedule;
+import org.bukkit.NamespacedKey;
+import org.bukkit.Registry;
+import org.jetbrains.annotations.NotNull;
+
+public class CraftEntitySchedule extends BukkitToNMSMapping<Schedule> implements EntitySchedule {
+    public static EntitySchedule minecraftToBukkit(Schedule minecraft) {
+        return minecraftToBukkit(minecraft, Registries.SCHEDULE, Registry.ENTITY_SCHEDULE);
+    }
+
+    public static Schedule bukkitToMinecraft(EntitySchedule bukkit) {
+        return bukkitToMinecraft(bukkit, Registries.SCHEDULE);
+    }
+
+    public CraftEntitySchedule(NamespacedKey key, Schedule handle) {
+        super(key, handle);
+    }
+
+    @Override
+    public @NotNull String translationKey() {
+        return "schedule.minecraft." + getKey().getKey();
+    }
+
+    @Override
+    public String toString() {
+        return "CraftEntitySchedule{key=" + this.getKey() + "}";
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftPoiType.java b/src/main/java/org/bukkit/craftbukkit/CraftPoiType.java
new file mode 100644
index 0000000000000000000000000000000000000000..5c6d277657dcc5aa7a03cceb87124fa8dfcc24c0
--- /dev/null
+++ b/src/main/java/org/bukkit/craftbukkit/CraftPoiType.java
@@ -0,0 +1,31 @@
+package org.bukkit.craftbukkit;
+
+import net.minecraft.core.registries.Registries;
+import org.bukkit.NamespacedKey;
+import org.bukkit.PoiType;
+import org.bukkit.Registry;
+import org.jetbrains.annotations.NotNull;
+
+public class CraftPoiType extends BukkitToNMSMapping<net.minecraft.world.entity.ai.village.poi.PoiType> implements PoiType {
+    public static PoiType minecraftToBukkit(net.minecraft.world.entity.ai.village.poi.PoiType minecraft) {
+        return minecraftToBukkit(minecraft, Registries.POINT_OF_INTEREST_TYPE, Registry.POI_TYPE);
+    }
+
+    public static net.minecraft.world.entity.ai.village.poi.PoiType bukkitToMinecraft(PoiType bukkit) {
+        return bukkitToMinecraft(bukkit, Registries.POINT_OF_INTEREST_TYPE);
+    }
+
+    public CraftPoiType(NamespacedKey key, net.minecraft.world.entity.ai.village.poi.PoiType handle) {
+        super(key, handle);
+    }
+
+    @Override
+    public @NotNull String translationKey() {
+        return "poitype.minecraft."+this.getKey().getKey();
+    }
+
+    @Override
+    public String toString() {
+        return "CraftPoiType{key=" + this.getKey() + "}";
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttribute.java b/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttribute.java
index 0d04f0a34d1d1894845b720a407f7190ea78d514..eef30b964e5739a5459b76593ceb1f24497c10c5 100644
--- a/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttribute.java
+++ b/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttribute.java
@@ -27,16 +27,7 @@ public class CraftAttribute implements Attribute, Handleable<net.minecraft.world
 
     public static Attribute stringToBukkit(String string) {
         Preconditions.checkArgument(string != null);
-
-        // We currently do not have any version-dependent remapping, so we can use current version
-        // First convert from when only the names where saved
-        string = FieldRename.convertAttributeName(ApiVersion.CURRENT, string);
-        string = string.toLowerCase(Locale.ROOT);
-        NamespacedKey key = NamespacedKey.fromString(string);
-        if (key == null) return null; // Paper - Fixup NamespacedKey handling
-
-        // Now also convert from when keys where saved
-        return CraftRegistry.get(Registry.ATTRIBUTE, key, ApiVersion.CURRENT);
+        return Registry.ATTRIBUTE.get(NamespacedKey.fromString(string));
     }
 
     public static net.minecraft.world.entity.ai.attributes.Attribute bukkitToMinecraft(Attribute bukkit) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index f246df4e117dddd08ae6314050ad3e73114d765b..a9be4b619f4e6f23b9d3ebeff7a405de7092e587 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -933,7 +933,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
                     continue;
                 }
                 AttributeModifier modifier = (AttributeModifier) o;
-                Attribute attribute = CraftAttribute.stringToBukkit(attributeName);
+                Attribute attribute = Attribute.getAttribute(org.bukkit.NamespacedKey.minecraft(attributeName.toUpperCase(Locale.ROOT))); // MCCLab
                 if (attribute == null) {
                     continue;
                 }
@@ -2328,7 +2328,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             if (modCollection == null || modCollection.isEmpty()) {
                 continue;
             }
-            mods.put(CraftAttribute.bukkitToString(entry.getKey()), new ArrayList<>(modCollection));
+            mods.put(entry.getKey().getKey().getKey().replace(".", "_"), new ArrayList<>(modCollection)); // MCCLab
         }
         builder.put(key.BUKKIT, mods);
     }
diff --git a/src/test/java/org/bukkit/support/provider/RegistriesArgumentProvider.java b/src/test/java/org/bukkit/support/provider/RegistriesArgumentProvider.java
index dc5fadb3d98b443df54b554168d60fe0b0226664..567eab3ac448f6e6fd9d2bdec245da13702f259a 100644
--- a/src/test/java/org/bukkit/support/provider/RegistriesArgumentProvider.java
+++ b/src/test/java/org/bukkit/support/provider/RegistriesArgumentProvider.java
@@ -14,6 +14,8 @@ import net.minecraft.world.entity.animal.WolfVariant;
 import net.minecraft.world.entity.decoration.PaintingVariant;
 import net.minecraft.world.entity.npc.VillagerProfession;
 import net.minecraft.world.entity.npc.VillagerType;
+import net.minecraft.world.entity.schedule.Activity;
+import net.minecraft.world.entity.schedule.Schedule;
 import net.minecraft.world.item.Instrument;
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.saveddata.maps.MapDecorationType;
@@ -25,6 +27,7 @@ import org.bukkit.MusicInstrument;
 import org.bukkit.Sound;
 import org.bukkit.attribute.Attribute;
 import org.bukkit.block.Biome;
+import org.bukkit.*;
 import org.bukkit.block.BlockType;
 import org.bukkit.block.banner.PatternType;
 import org.bukkit.craftbukkit.CraftArt;
@@ -35,6 +38,7 @@ import org.bukkit.craftbukkit.CraftMusicInstrument;
 import org.bukkit.craftbukkit.CraftSound;
 import org.bukkit.craftbukkit.attribute.CraftAttribute;
 import org.bukkit.craftbukkit.block.CraftBiome;
+import org.bukkit.craftbukkit.*;
 import org.bukkit.craftbukkit.block.CraftBlockType;
 import org.bukkit.craftbukkit.block.banner.CraftPatternType;
 import org.bukkit.craftbukkit.damage.CraftDamageType;
@@ -101,6 +105,10 @@ public class RegistriesArgumentProvider implements ArgumentsProvider {
         register(RegistryKey.BANNER_PATTERN, PatternType.class, Registries.BANNER_PATTERN, CraftPatternType.class, BannerPattern.class);
         register(RegistryKey.MENU, MenuType.class, Registries.MENU, CraftMenuType.class, net.minecraft.world.inventory.MenuType.class);
         register(RegistryKey.DATA_COMPONENT_TYPE, io.papermc.paper.datacomponent.DataComponentType.class, Registries.DATA_COMPONENT_TYPE, io.papermc.paper.datacomponent.PaperDataComponentType.class, net.minecraft.core.component.DataComponentType.class);
+
+        register(RegistryKey.POI_TYPE, PoiType.class, Registries.POINT_OF_INTEREST_TYPE, CraftPoiType.class, net.minecraft.world.entity.ai.village.poi.PoiType.class);
+        register(RegistryKey.ENTITY_ACTIVITY, EntityActivity.class, Registries.ACTIVITY, CraftEntityActivity.class, Activity.class);
+        register(RegistryKey.ENTITY_SCHEDULE, EntitySchedule.class, Registries.SCHEDULE, CraftEntitySchedule.class, Schedule.class);
     }
 
     private static void register(RegistryKey registryKey, Class bukkit, ResourceKey registry, Class craft, Class minecraft) { // Paper
