From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Fri, 28 Jun 2024 14:05:05 +0000
Subject: [PATCH] FurnaceRunOutOfFuelEvent called correctly now


diff --git a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index 9c89397a838d904f45b735c903d8ed16c29363aa..94732a36ab28b79f288a67122a395b009cd715f7 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -306,7 +306,9 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
 
         if (flag1) {
             setChanged(world, pos, state);
-            if(!flag)
+            if(flag)
+/*                new org.bukkit.event.inventory.FurnaceRunOutOfFuelEvent(CraftBlock.at(world, pos)).callEvent();
+            else*/
                 new org.bukkit.event.inventory.FurnaceRunOutOfFuelEvent(CraftBlock.at(world, pos)).callEvent();
         }
 
