From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: verdox <mail.ysp@web.de>
Date: Thu, 22 Aug 2024 21:18:24 +0000
Subject: [PATCH] Add WorldSelectChunkGeneratorEvent


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6fc8288fd4820f90f2713d860edb58f025caff24..31ef4b67c8e46b068d00b0a07fb6cf66f80e717c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1965,7 +1965,9 @@ public final class CraftServer implements Server {
             }
         }
 
-        return result;
+        io.papermc.paper.event.world.WorldSelectChunkGeneratorEvent worldSelectChunkGeneratorEvent = new io.papermc.paper.event.world.WorldSelectChunkGeneratorEvent(world, result);
+        worldSelectChunkGeneratorEvent.callEvent();
+        return worldSelectChunkGeneratorEvent.getPickedWorldGenerator();
     }
 
     public BiomeProvider getBiomeProvider(String world) {
